### PR TITLE
Heedls 499 deactivate delegate

### PR DESCRIPTION
**Pull request by JBsoftwire**
### JIRA link
https://softwiretech.atlassian.net/browse/HEEDLS-499

### Description
The deactivate button on the view delegate page now deactivates the delegate account in question and refreshes the page.

### Screenshots
No visual changes.

-----
### Developer checks
I have:
- [X] Run the formatter and made sure there are no IDE errors.
- [X] Written tests for the changes (controller, data services, services, view models etc) and manually tested my work with and without JavaScript. Full manual testing guidelines can be found here: https://softwiretech.atlassian.net/wiki/spaces/HEE/pages/6703648740/Testing
- [] Updated/added documentation in Swiki and/or Readme.
- [] Updated my Jira ticket with information about other parts of the system that were touched as part of the MR and have to be sanity tested to ensure nothing’s broken.
- [X] Scanned over my own MR to ensure everything is as expected.

### CODE REVIEW

File: DigitalLearningSolutions.Data/DataServices/UserDataService/UserDataService.cs [hunk 1]
```diff
@@ -86,6 +86,8 @@ string emailAddress
 
         int GetDelegateCountWithAnswerForPrompt(int centreId, int promptNumber);
         void DeleteAllAnswersForPrompt(int centreId, int promptNumber);
+
+        void DeactivateDelegateUser(int delegateId);
```
Comment (AlexJacksonDS) [inline DigitalLearningSolutions.Data/DataServices/UserDataService/UserDataService.cs:90]: Can you move this up next to the other methods in the DelegateUserDataService file?

File: DigitalLearningSolutions.Web/Views/TrackingSystem/Delegates/ViewDelegate/Index.cshtml [hunk 2]
```diff
@@ -109,7 +109,7 @@
           Promote to admin
         </a>
         <div>
-          <a class="nhsuk-button delete-button view-delegate-button" role="button">
+          <a class="nhsuk-button delete-button view-delegate-button" role="button" asp-controller="ViewDelegate" asp-action="DeactivateDelegate" asp-route-delegateId="@Model.DelegateInfo.Id">
```
Comment (AlexJacksonDS) [inline DigitalLearningSolutions.Web/Views/TrackingSystem/Delegates/ViewDelegate/Index.cshtml:112]: This line is quite long, can you break it down.

**Review by AlexJacksonDS — CHANGES_REQUESTED** [review]
Just a couple of minor formatting issues.

**Review by AlexJacksonDS — APPROVED** [review]

File: DigitalLearningSolutions.Web/Controllers/TrackingSystem/Delegates/ViewDelegateController.cs [hunk 3]
```diff
@@ -68,5 +68,20 @@ public IActionResult SendWelcomeEmail(int delegateId)
 
             return View("WelcomeEmailSent", model);
         }
+
+        [Route("DeactivateDelegate")]
+        public IActionResult DeactivateDelegate(int delegateId)
```
Comment (stellake) [inline DigitalLearningSolutions.Web/Controllers/TrackingSystem/Delegates/ViewDelegateController.cs:73]: Given that we don't have a confirmation page for the delegates, should we make this POST?

Also, should we add a few controller test (no delegate user, wrong centre and default cases)?

File: DigitalLearningSolutions.Web.Tests/Controllers/TrackingSystem/Delegates/ViewDelegateControllerTests.cs [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+            userDataService = A.Fake<IUserDataService>();
+            var courseService = A.Fake<ICourseService>();
+            var passwordResetService = A.Fake<IPasswordResetService>();
+
+            var httpRequest = A.Fake<HttpRequest>();
+            var httpResponse = A.Fake<HttpResponse>();
+            const string cookieName = "";
+            const string cookieValue = "";
+
+            viewDelegateController = new ViewDelegateController(userDataService, centreCustomPromptsHelper, courseService, passwordResetService)
```
Comment (stellake) [inline DigitalLearningSolutions.Web.Tests/Controllers/TrackingSystem/Delegates/ViewDelegateControllerTests.cs:34]: Nice tests! Do we need all this setup though (e.g. the MockTempData, and could HTTPContext potentially be defaultcontext)?

File: DigitalLearningSolutions.Web/Controllers/TrackingSystem/Delegates/ViewDelegateController.cs [hunk 5]
```diff
@@ -69,6 +69,7 @@ public IActionResult SendWelcomeEmail(int delegateId)
             return View("WelcomeEmailSent", model);
         }
 
+        [HttpPost]
```
Comment (stellake) [inline DigitalLearningSolutions.Web/Controllers/TrackingSystem/Delegates/ViewDelegateController.cs:72]: Do we also need to change the button in the view?

**Review by stellake — APPROVED** [review]
Looks good! 👍

**Pull request merged**